File: src/oca/java/src/org/opennebula/client/document/Document.java
/*******************************************************************************
 * Copyright 2002-2021, OpenNebula Project, OpenNebula Systems
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 ******************************************************************************/
package org.opennebula.client.document;

import org.opennebula.client.Client;
import org.opennebula.client.OneResponse;
import org.opennebula.client.PoolElement;
import org.w3c.dom.Node;

/**
 * This class represents an OpenNebula document.
 * Documents are generic objects. You can dynamically create new Pools in
 * OpenNebula, creating subclasses with different TYPE values.
 * <br>
 * TYPE must be the same for the corresponding pool, see {@link DocumentPool}
 * <br>
 * For example:
 * <pre>
 * <code>
 * public class GenericObjA extends Document
 * {
 *     private static final int TYPE = 200;
 *
 *     public GenericObjA(int id, Client client)
 *     {
 *         super(id, client);
 *     }
 *
 *     public GenericObjA(Node xmlElement, Client client)
 *     {
 *         super(xmlElement, client);
 *     }
 *
 *     public static OneResponse allocate(Client client, String description)
 *     {
 *         return Document.allocate(client, description, TYPE);
 *     }
 * }
 * </code>
 * </pre>
 *
 */
public abstract class Document extends PoolElement
{

    private static final String METHOD_PREFIX = "document.";
    private static final String ALLOCATE = METHOD_PREFIX + "allocate";
    private static final String DELETE   = METHOD_PREFIX + "delete";
    private static final String INFO     = METHOD_PREFIX + "info";

    private static final String UPDATE   = METHOD_PREFIX + "update";
    private static final String CHOWN    = METHOD_PREFIX + "chown";
    private static final String CHMOD    = METHOD_PREFIX + "chmod";
    private static final String CLONE    = METHOD_PREFIX + "clone";
    private static final String RENAME   = METHOD_PREFIX + "rename";
    private static final String LOCK     = METHOD_PREFIX + "lock";
    private static final String UNLOCK   = METHOD_PREFIX + "unlock";

    /**
     * Creates a new Document representation.
     * @param id The document id.
     * @param client XML-RPC Client.
     */
    public Document(int id, Client client)
    {
        super(id, client);
    }

    /**
     * @see PoolElement
     */
    public Document(Node xmlElement, Client client)
    {
        super(xmlElement, client);
    }

    /**
     * lock this Document
     *
     * @param client XML-RPC Client.
     * @param id The Image id.
     * @param level Lock level.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse lock(Client client, int id, int level)
    {
        return client.call(LOCK, id, level);
    }

    /**
     * Unlock this Document
     *
     * @param client XML-RPC Client.
     * @param id The Image id.
     * @return If an error occurs the error message contains the reason.
     */
    public static OneResponse unlock(Client client, int id)
    {
        return client.call(UNLOCK, id);
    }

    // =================================
    // Static XML-RPC methods
    // =================================

    /**
     * Allocates a new Document in OpenNebula.
     *
     * @param client XML-RPC Client.
     * @param description A string containing the template of the document.
     * @return If successful the message contains the associated
     * id generated for this Document.
     */
    protected static OneResponse allocate(Client client, String description, int type)
    {
        return client.call(ALLOCATE, description, type);
    }

    /**
     * Retrieves the information of the given Document.
     *
     * @return If successful the message contains the string
     * with the information returned by OpenNebula.
     */
    public OneResponse info()
    {
        return client.call(INFO, id);
    }

    /**
     * Deletes a document from OpenNebula.
     *
     * @return A encapsulated response.
     */
    public OneResponse delete()
    {
        return client.call(DELETE, id);
    }

    /**
     * Replaces the document template contents.
     *
     * @param new_document New template contents.
     * @return If successful the message contains the document id.
     */
    public OneResponse update(String new_document)
    {
        return update(new_document, false);
    }

    /**
     * Replaces the document template contents.
     *
     * @param new_document New template contents.
     * @param append True to append new attributes instead of replace the whole template
     * @return If successful the message contains the document id.
     */
    public OneResponse update(String new_document, boolean append)
    {
        return client.call(UPDATE, id, new_document, append);
    }

    /**
     * Publishes or unpublishes a document.
     *
     * @param publish True for publishing, false for unpublishing.
     * @return If successful the message contains the document id.
     */
    public OneResponse publish(boolean publish)
    {
        int group_u = publish ? 1 : 0;

        return chmod(-1, -1, -1, group_u, -1, -1, -1, -1, -1);
    }

    /**
     * Changes the owner/group
     *
     * @param uid The new owner user ID. Set it to -1 to leave the current one.
     * @param gid The new group ID. Set it to -1 to leave the current one.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid, int gid)
    {
        return client.call(CHOWN, id, uid, gid);
    }

    /**
     * Changes the document permissions
     *
     * @param owner_u 1 to allow, 0 deny, -1 do not change
     * @param owner_m 1 to allow, 0 deny, -1 do not change
     * @param owner_a 1 to allow, 0 deny, -1 do not change
     * @param group_u 1 to allow, 0 deny, -1 do not change
     * @param group_m 1 to allow, 0 deny, -1 do not change
     * @param group_a 1 to allow, 0 deny, -1 do not change
     * @param other_u 1 to allow, 0 deny, -1 do not change
     * @param other_m 1 to allow, 0 deny, -1 do not change
     * @param other_a 1 to allow, 0 deny, -1 do not change
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(   int owner_u, int owner_m, int owner_a,
                                int group_u, int group_m, int group_a,
                                int other_u, int other_m, int other_a)
    {
        return chmod(client, CHMOD, id,
                    owner_u, owner_m, owner_a,
                    group_u, group_m, group_a,
                    other_u, other_m, other_a);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(String octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Changes the permissions
     *
     * @param octet Permissions octed , e.g. 640
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chmod(int octet)
    {
        return chmod(client, CHMOD, id, octet);
    }

    /**
     * Clones this document into a new one
     *
     * @param name Name for the new document.
     * @return If successful the message contains the new document ID.
     */
    public OneResponse clone(String name)
    {
        return client.call(CLONE, id, name);
    }

    /**
     * Renames this document
     *
     * @param name New name for the document.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse rename(String name)
    {
        return client.call(RENAME, id, name);
    }

    /**
     * Locks this object
     *
     * @param level int to identify the lock level
     * @return In case of success, a boolean with true if the lock was granted,
     * and false if the object is already locked.
     */
    public OneResponse lock(int level)
    {
        return client.call(LOCK, id, level);
    }

    /**
     * Unlocks this object
     *
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse unlock()
    {
        return client.call(UNLOCK, id);
    }

    // =================================
    // Helpers
    // =================================

    /**
     * Publishes the document.
     *
     * @return If successful the message contains the document id.
     */
    public OneResponse publish()
    {
        return publish(true);
    }

    /**
     * Unpublishes the document.
     *
     * @return If successful the message contains the document id.
     */
    public OneResponse unpublish()
    {
        return publish(false);
    }

    /**
     * Changes the owner
     *
     * @param uid The new owner user ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chown(int uid)
    {
        return chown(uid, -1);
    }

    /**
     * Changes the group
     *
     * @param gid The new group ID.
     * @return If an error occurs the error message contains the reason.
     */
    public OneResponse chgrp(int gid)
    {
        return chown(-1, gid);
    }
}

File: src/oca/ruby/opennebula/document.rb
# -------------------------------------------------------------------------- #
# Copyright 2002-2024, OpenNebula Project, OpenNebula Systems                #
#                                                                            #
# Licensed under the Apache License, Version 2.0 (the "License"); you may    #
# not use this file except in compliance with the License. You may obtain    #
# a copy of the License at                                                   #
#                                                                            #
# http://www.apache.org/licenses/LICENSE-2.0                                 #
#                                                                            #
# Unless required by applicable law or agreed to in writing, software        #
# distributed under the License is distributed on an "AS IS" BASIS,          #
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.   #
# See the License for the specific language governing permissions and        #
# limitations under the License.                                             #
#--------------------------------------------------------------------------- #

require 'opennebula/lockable_ext'
require 'opennebula/pool_element'

module OpenNebula

    # All subclasses must define the DOCUMENT_TYPE constant.
    #
    # @example
    #   require 'opennebula/document'
    #
    #   module OpenNebula
    #       class CustomObject < Document
    #
    #           DOCUMENT_TYPE = 400
    #
    #       end
    #   end
    class Document < PoolElement

        #######################################################################
        # Constants and Class Methods
        #######################################################################

        DOCUMENT_METHODS = {
            :allocate   => 'document.allocate',
            :delete     => 'document.delete',
            :info       => 'document.info',
            :update     => 'document.update',
            :chown      => 'document.chown',
            :chmod      => 'document.chmod',
            :clone      => 'document.clone',
            :rename     => 'document.rename',
            :lock       => 'document.lock',
            :unlock     => 'document.unlock'
        }

        # Creates a Document Object description with just its identifier
        # this method should be used to create plain Document objects.
        # @param [Integer] pe_id the id of the object
        #
        # @return [Nokogiri::XML::Node, REXML::Element] the empty xml
        def self.build_xml(pe_id = nil)
            if pe_id
                obj_xml = "<DOCUMENT><ID>#{pe_id}</ID></DOCUMENT>"
            else
                obj_xml = '<DOCUMENT></DOCUMENT>'
            end

            XMLElement.build_xml(obj_xml, 'DOCUMENT')
        end

        # Class constructor
        #
        # @param [Nokogiri::XML::Node, REXML::Element] xml string
        #   created by the build_xml() method
        # @param [OpenNebula::Client] client the xml-rpc client
        #
        # @return [Document] the new object
        #
        # @example
        #   doc = Document.new(Document.build_xml(3),rpc_client)
        def initialize(xml, client)
            LockableExt.make_lockable(self, DOCUMENT_METHODS)

            super(xml, client)
        end

        #######################################################################
        # XML-RPC Methods for the Document Object
        #######################################################################

        # Retrieves the information of the given Document.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def info(decrypt = false)
            rc = super(DOCUMENT_METHODS[:info], 'DOCUMENT', decrypt)

            if !OpenNebula.is_error?(rc) && self['TYPE'].to_i != document_type
                return OpenNebula::Error.new("[DocumentInfo] Error getting document [#{@pe_id}].")
            end

            return rc
        end

        alias info! info

        # Allocates a new Document in OpenNebula
        #
        # @param description [String] The contents of the Document.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def allocate(description)
            super(DOCUMENT_METHODS[:allocate], description, document_type)
        end

        alias allocate_xml allocate

        # Deletes the Document
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def delete
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            return call(DOCUMENT_METHODS[:delete], @pe_id)
        end

        # Replaces the template contents
        #
        # @param [String] new_template new template contents
        # @param append [true, false] True to append new attributes instead of
        #   replace the whole template
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update(new_template, append = false)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            super(DOCUMENT_METHODS[:update], new_template, append ? 1 : 0)
        end

        # Replaces the raw template contents
        #
        # @param template [String] New template contents, in the form KEY = VAL
        # @param append [true, false] True to append new attributes instead of
        #   replace the whole template
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def update_raw(template_raw, append = false)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            return call(DOCUMENT_METHODS[:update], @pe_id, template_raw, append ? 1 : 0)
        end

        # Changes the owner/group
        #
        # @param [Integer] uid the new owner id. Set to -1 to leave the current one
        # @param [Integer] gid the new group id. Set to -1 to leave the current one
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chown(uid, gid)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            super(DOCUMENT_METHODS[:chown], uid, gid)
        end

        # Changes the Document permissions.
        #
        # @param octet [String] Permissions octed , e.g. 640
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod_octet(octet)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            super(DOCUMENT_METHODS[:chmod], octet)
        end

        # Changes the Document permissions.
        # Each [Integer] argument must be 1 to allow, 0 deny, -1 do not change
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def chmod(owner_u, owner_m, owner_a, group_u, group_m, group_a, other_u,
                  other_m, other_a)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            super(DOCUMENT_METHODS[:chmod], owner_u, owner_m, owner_a, group_u,
                group_m, group_a, other_u, other_m, other_a)
        end

        # Clones this Document into a new one
        #
        # @param name [String] Name for the new Document.
        #
        # @return [Integer, OpenNebula::Error] The new Document ID in case
        #   of success, Error otherwise
        def clone(name)
            rc = check_type
            return rc if OpenNebula.is_error?(rc)

            return Error.new('ID not defined') unless @pe_id

            rc = @client.call(DOCUMENT_METHODS[:clone], @pe_id, name)

            return rc
        end

        # Renames this Document
        #
        # @param name [String] New name for the Document.
        #
        # @return [nil, OpenNebula::Error] nil in case of success, Error
        #   otherwise
        def rename(name)
            return call(DOCUMENT_METHODS[:rename], @pe_id, name)
        end

        #######################################################################
        # Helpers to get Document information
        #######################################################################

        # Returns the group identifier
        # @return [Integer] the element's group ID
        def gid
            self['GID'].to_i
        end

        # Returns the owner user ID
        # @return [Integer] the element's owner user ID
        def owner_id
            self['UID'].to_i
        end

        # Returns true if the GROUP_U permission bit is set
        # @return [true, false] true if the GROUP_U permission bit is set
        def public?
            if self['PERMISSIONS/GROUP_U'] == '1' || self['PERMISSIONS/OTHER_U'] == '1'
                true
            else
                false
            end
        end

        def document_type
            self.class::DOCUMENT_TYPE
        end

        private

        def set_publish(published)
            group_u = published ? 1 : 0

            chmod(-1, -1, -1, group_u, -1, -1, -1, -1, -1)
        end

        def check_type
            type = self['TYPE']

            if type.nil? && @pe_id
                rc = @client.call(DOCUMENT_METHODS[:info], @pe_id)

                return rc if OpenNebula.is_error?(rc)

                xmldoc = XMLElement.new
                xmldoc.initialize_xml(rc, 'DOCUMENT')

                type = xmldoc['TYPE']
            end

            if !type.nil? && type.to_i != document_type
                return OpenNebula::Error.new(
                    "[DocumentInfo] Error getting document [#{@pe_id}]."
                )
            end

            return
        end

    end

end
